File: app/controllers/batch/recognition.py
"""Batch Recognition Controller"""
from datetime import datetime

from app.services.batch import BatchRecognitionService
from app.schemas import BatchRecognitionResult, BatchRecognitionResponse, FaceLocation
from app.utils.exceptions import ResourceNotFoundException


class BatchRecognitionController:
    """Controller for batch recognition operations."""
    
    def __init__(self, service: BatchRecognitionService):
        self.service = service
    
    def process_image(self, image_path: str) -> BatchRecognitionResponse:
        """Process an image for face recognition."""
        result = self.service.process_image(image_path)
        
        recognition_results = [
            BatchRecognitionResult(
                _id=r.get("_id"),
                person_name=r["person_name"],
                detection_datetime=datetime.utcnow(),
                confidence_score=r["confidence_score"],
                face_location=FaceLocation(
                    x_min=r["face_location"]["x_min"],
                    y_min=r["face_location"]["y_min"],
                    x_max=r["face_location"]["x_max"],
                    y_max=r["face_location"]["y_max"]
                )
            )
            for r in result["results"]
        ]
        
        return BatchRecognitionResponse(
            success=True,
            batch_id=result["batch_id"],
            source_path=result["source_path"],
            total_faces_detected=result["total_faces_detected"],
            recognized_faces=result["recognized_faces"],
            processing_time_ms=result["processing_time_ms"],
            results=recognition_results
        )
    
    def get_batch_results(self, batch_id: str) -> dict:
        """Get results for a specific batch."""
        result = self.service.get_batch_results(batch_id)
        if not result:
            raise ResourceNotFoundException(f"Batch {batch_id} not found")
        
        return {
            "batch_id": result["batch_id"],
            "source_path": result["source_path"],
            "total_faces_detected": result["total_faces_detected"],
            "recognized_count": result["recognized_count"],
            "processing_time_ms": result["processing_time_ms"],
            "results": [BatchRecognitionResult.from_mongo(log) for log in result.get("logs", [])]
        }
    
    def get_all_batches(self, skip: int = 0, limit: int = 10) -> dict:
        """Get all batch summaries."""
        batches = self.service.get_all_batches(skip, limit)
        return {
            "skip": skip, "limit": limit,
            "batches": [
                {
                    "batch_id": b["_id"],
                    "source_path": b["source_path"],
                    "detection_datetime": b["detection_datetime"],
                    "total_faces_detected": b["total_faces_detected"],
                    "recognized_count": b["recognized_count"],
                    "processing_time_ms": b["processing_time_ms"]
                }
                for b in batches
            ]
        }
    
    def delete_all_batches(self) -> dict:
        """Delete all batch recognition logs."""
        return self.service.delete_all_batches()

File: frontend/src/api/recognition.ts
import apiClient from './client';

// Types
export interface RecognitionLog {
    _id: string;
    person_name: string;
    detection_datetime: string;
    confidence_score: number;
    camera_id?: string;
}

export interface BatchResult {
    _id: string;
    image_path: string;
    total_faces: number;
    recognized_count: number;
    unknown_count: number;
    results: RecognitionLog[];
}

export interface Person {
    name: string;
    image_count: number;
}

// Group Recognition Types
export interface RecognizedStudent {
    student_id: string;
    admission_id: string;
    first_name: string;
    last_name: string;
    course_name: string;
    section: string;
    confidence: number;
    face_location: {
        x_min: number;
        y_min: number;
        x_max: number;
        y_max: number;
    };
}

export interface GroupRecognitionResult {
    success: boolean;
    batch_id: string;
    stats: {
        total_faces_detected: number;
        recognized_count: number;
        unknown_count: number;
        processing_time_ms: number;
    };
    recognized: RecognizedStudent[];
    unknown_faces: Array<{ face_location: { note?: string } }>;
}

// Get recognition logs
export const getLogs = async (skip = 0, limit = 10) => {
    const response = await apiClient.get('/recognition/logs', { params: { skip, limit } });
    return response.data;
};

// Process batch image
export const processBatch = async (imagePath: string) => {
    const response = await apiClient.post('/recognition/batch/process', { image_path: imagePath });
    return response.data;
};

// Get all batches
export const getBatches = async (skip = 0, limit = 10) => {
    const response = await apiClient.get('/recognition/batches', { params: { skip, limit } });
    return response.data;
};

// Get known persons
export const getPersons = async () => {
    const response = await apiClient.get('/admission/known-persons');
    return response.data;
};

// Add persons from folder
export const addPersons = async (sourcePath?: string) => {
    const response = await apiClient.post('/recognition/persons/add', { source_path: sourcePath });
    return response.data;
};

// Recognize students from group photo
export const recognizeGroup = async (imageFile: File): Promise<GroupRecognitionResult> => {
    const formData = new FormData();
    formData.append('image', imageFile);
    
    const response = await apiClient.post('/admission/recognize-group', formData, {
        headers: {
            'Content-Type': 'multipart/form-data',
        },
    });
    return response.data;
};

// History log type
export interface RecognitionHistoryLog {
    id: string;
    student_id: string;
    admission_id: string;
    first_name: string;
    last_name: string;
    course_name: string;
    section: string;
    confidence: number;
    detection_datetime: string;
    batch_id: string;
}

export interface RecognitionHistoryResult {
    success: boolean;
    total: number;
    skip: number;
    limit: number;
    logs: RecognitionHistoryLog[];
}

// Get recognition history with filters
export const getRecognitionHistory = async (params: {
    student_id?: string;
    course_id?: string;
    start_date?: string;
    end_date?: string;
    skip?: number;
    limit?: number;
}): Promise<RecognitionHistoryResult> => {
    const response = await apiClient.get('/admission/recognition-history', { params });
    return response.data;
};


